scripts/test_metric_reporter/suite_reporter.py

```diff
@@ -0,0 +1,186 @@
+        """Calculate the percentage rate of a given value over the total.
+
+        Args:
+            value (int): The numerator for the rate calculation.
+            total (int): The denominator for the rate calculation.
+
+        Returns:
+            float | None: The calculated rate as a percentage, or None if the total is 0.
+        """
+        return (value / total) * 100 if total > 0 else None
+
+    def dict_with_fieldnames(self) -> dict[str, str | int | float | None]:
+        """Convert the test suite result to a dictionary with field names.
+
+        Returns:
+            dict[str, str | int | float | None]: Dictionary representation of the test suite result.
+        """
+        return {
+            "Date - Job Number": self.date_job,
+            "Date": self.date,
+            "Job Number": self.job,
+            "Status": self.status,
+            "Execution Time": self.execution_time,
+            "Success": self.success,
+            "Failure": self.failure,
+            "Skipped": self.skipped,
+            "Unknown": self.unknown,
+            "Total": self.total,
+            "Success Rate (%)": self.success_rate,
+            "Failure Rate (%)": self.failure_rate,
+            "Skip Rate (%)": self.skipped_rate,
+            "Unknown Rate (%)": self.unknown_rate,
+        }
+
+
+class SuiteReporterError(BaseError):
+    """Exception raised for errors in the SuiteReporter."""
+
+    pass
+
+
+class SuiteReporter:
+    """Handles the reporting of test suite results from CircleCI metadata."""
+
+    logger = logging.getLogger(__name__)
+
+    def __init__(self, test_metadata_directory: str) -> None:
+        """Initialize the reporter with the directory containing test metadata.
+
+        Args:
+            test_metadata_directory (str): Path to the directory with CircleCI test metadata.
+
+        Raises:
+            CircleCIJsonParserError: If there is an error parsing the test metadata.
+        """
+        self._test_metadata_directory = test_metadata_directory
+        self.results: list[SuiteReporterResult] = self._parse_circleci_metadata()
+
+    def _parse_circleci_metadata(self) -> list[SuiteReporterResult]:
+        results: list[SuiteReporterResult] = []
+        for job_test_metadata in CircleCIJsonParser.parse(self._test_metadata_directory):
+            started_at = datetime.strptime(job_test_metadata.job.started_at, DATETIME_FORMAT)
+            stopped_at = datetime.strptime(job_test_metadata.job.stopped_at, DATETIME_FORMAT)
+            execution_time = (stopped_at - started_at).total_seconds()
+            test_suite_result = SuiteReporterResult(
+                date=started_at.strftime(DATE_FORMAT),
+                job=job_test_metadata.job.job_number,
+                status=job_test_metadata.job.status,
+                execution_time=execution_time,
+            )
+
+            if not job_test_metadata.test_metadata:
+                continue
```

Comment:
Q: should this be moved up to L128.5?

Not sure if I'm reading the code correctly, but it seems like we might be juggling L129-137 and then discarding those changes if `job_test_metadata.test_metadata` is falsey, and the results might only be appended below on L151.

Reply:
You are right we can do it right at the start to be more efficient (see https://github.com/mozilla/ecosystem-test-scripts/pull/4/commits/6390587d42f3df9f76f7d35b9c808d25dec2cee8)